File: crates/obscura-js/js/bootstrap/tools/trace-source.js
// Execution-source attribution for the trace streams: the `from` label that
// says which code unit produced a traced record. The taxonomy mirrors the
// HaHaVM dispatch trace (window / iframe(N) / worker(M)[creator] /
// script@<url> / function@<source> / eval@<source> / host) so traces from the
// two engines can be diffed unit by unit.
//
// The authoritative stack lives in Rust (thread-local, read by the host-op
// trace writer); this module mirrors it for the snapshots JS needs (a timer
// registered by script A must restore A's label when it fires, not inherit
// whoever ran last) and keeps the two in lockstep through
// op_trace_push_source / op_trace_pop_source. Rust funnels set the ambient
// label per context; the host sets `__obscura_trace_default_from` before this
// realm's bootstrap runs.
//
// Production runs (no trace env var) keep every wrapper off: the flag global
// is never set, enter/leave are no-ops, `Function` and `Promise.prototype.then`
// stay native, and on* handlers stay plain data properties. The only residual
// cost is two global function calls around each Rust-executed script.
(function _installTraceSource() {
  const G = globalThis;
  const enabled = () => G.__obscura_trace_from_enabled === true;

  // JS-side mirror of the Rust label stack. `current === null` means "ambient"
  // (window / iframe(N) / worker(M)[...] per realm).
  let stack = [];
  let current = null;

  function defaultFrom() {
    const d = G.__obscura_trace_default_from;
    return typeof d === 'string' && d ? d : 'window';
  }
  function syncRust(label) {
    try {
      Deno.core.ops.op_trace_push_source(label);
    } catch (e) { /* tracing stays best-effort */ }
  }
  function desyncRust() {
    try {
      Deno.core.ops.op_trace_pop_source();
    } catch (e) { /* tracing stays best-effort */ }
  }
  // A hidden own data string the V8 JSONL writer can read without calling
  // back into JS. Updating it here keeps the IC path off the JS stack.
  function publishFrom() {
    if (!enabled()) return;
    const label = current || defaultFrom();
    try {
      Object.defineProperty(G, '__obscura_trace_from', {
        value: label, writable: true, enumerable: false, configurable: true,
      });
    } catch (e) {
      try { G.__obscura_trace_from = label; } catch (e2) {}
    }
  }

  G.__obscuraTraceCurrent = function () {
    return current || defaultFrom();
  };
  // Rust-driven script brackets (page.rs) enter/leave around each classic
  // script; JS-driven ones (dynamic insertion, string timers) call the same
  // pair from __runClassicScript.
  G.__obscuraTraceEnter = function (label) {
    if (!enabled()) return;
    stack.push(current);
    current = String(label);
    syncRust(current);
    publishFrom();
  };
  G.__obscuraTraceLeave = function () {
    if (!enabled()) return;
    current = stack.length ? stack.pop() : null;
    desyncRust();
    publishFrom();
  };

  // Snapshot-at-registration / restore-at-invocation for async continuations.
  // Returns the function untouched unless tracing is on.
  G.__obscuraTraceBind = function (fn) {
    if (!enabled() || typeof fn !== 'function') return fn;
    const label = current || defaultFrom();
    return function (...args) {
      const previous = current;
      const depth = stack.length;
      current = label;
      syncRust(label);
      publishFrom();
      try {
        return fn.apply(this, args);
      } finally {
        // Force-unwind anything the callback leaked (a script terminated
        // mid-run by the watchdog leaves its entry behind).
        stack.length = depth;
        current = previous;
        desyncRust();
        publishFrom();
      }
    };
  };
  // Invoke one callback under a snapshot label (listener entries, on*
  // handlers). A null label means "no snapshot; run under the ambient turn".
  G.__obscuraTraceCallWith = function (label, fn, thisArg, args) {
    if (label === null || label === undefined || !enabled()) {
      return fn.apply(thisArg, args);
    }
    const previous = current;
    const depth = stack.length;
    current = String(label);
    syncRust(current);
    publishFrom();
    try {
      return fn.apply(thisArg, args);
    } finally {
      stack.length = depth;
      current = previous;
      desyncRust();
      publishFrom();
    }
  };

  // on* property-handler slots. Chrome exposes every event-handler IDL
  // attribute as a prototype (or Window own) accessor pair, so assignment
  // never creates an own property on the instance and `delete el.onclick`
  // is a no-op. That is the production shape here too; under trace the
  // setter additionally snapshots the assigning code's label, so a handler
  // is attributed to the code that assigned it rather than to whichever
  // turn dispatched the event.
  const handlerSlots = new WeakMap();
  function handlerSlot(target) {
    let slot = handlerSlots.get(target);
    if (!slot) {
      slot = new Map();
      handlerSlots.set(target, slot);
    }
    return slot;
  }
  G.__obscuraTraceDefineHandler = function (target, name, enumerable) {
    // The descriptor is installed on a prototype (or on Window itself).
    // Storage keys on `this`: Chrome keeps one handler value per instance,
    // and using `target` here would make every Element share one onclick.
    Object.defineProperty(target, name, {
      get() {
        const entry = handlerSlots.get(this)?.get(name);
        return entry ? entry.value : null;
      },
      set(value) {
        handlerSlot(this).set(name, {
          value: (typeof value === 'function' || (value && typeof value === 'object'))
            ? value : null,
          from: enabled() ? (current || defaultFrom()) : null,
        });
      },
      enumerable: !!enumerable,
      configurable: true,
    });
    return true;
  };
  // Record a snapshot from an existing setter (Image own accessors, the
  // Worker prototype handlers): the value itself stays where that setter
  // already put it.
  G.__obscuraTraceRecordHandler = function (target, name) {
    if (!enabled()) return;
    handlerSlot(target).set(name, { value: null, from: current || defaultFrom() });
  };
  G.__obscuraTraceHandlerFrom = function (target, name) {
    if (!enabled()) return null;
    const entry = handlerSlots.get(target)?.get(name);
    return entry ? entry.from : null;
  };

  // `new Function` products run as one labeled unit (function@<creation
  // source>). A Proxy over the original constructor keeps instanceof,
  // .prototype and Reflect.construct semantics intact; the only observable
  // difference under trace is Function.prototype.toString on a wrapped
  // product, which reads as [native code] -- the same surface the HaHaVM
  // reference build accepts.
  G.__obscuraTraceRawFunction = Function;
  if (enabled()) {
    const OriginalFunction = Function;
    const wrapProduct = (product) => {
      if (typeof product !== 'function') return product;
      const label = 'function@' + (current || defaultFrom());
      return new Proxy(product, {
        apply(target, thisArg, args) {
          return G.__obscuraTraceCallWith(label, target, thisArg, args);
        },
        construct(target, args) {
          return G.__obscuraTraceCallWith(label, () => Reflect.construct(target, args));
        },
      });
    };
    G.Function = new Proxy(OriginalFunction, {
      apply(target, _thisArg, args) {
        return wrapProduct(Reflect.apply(target, undefined, args));
      },
      construct(target, args) {
        return wrapProduct(Reflect.construct(target, args));
      },
    });
    // Keep the constructor identity Chrome guarantees: in a real browser
    // `f.constructor === Function` and `Function.prototype.constructor ===
    // Function` are the same object as the global. The proxy replaced the
    // global binding, so point the prototype's constructor at it too.
    try {
      Object.defineProperty(OriginalFunction.prototype, 'constructor', {
        value: G.Function,
        writable: true,
        enumerable: false,
        configurable: true,
      });
    } catch (e) { /* identity fix is best-effort */ }
  }

  // Promise continuations carry the .then call site's label. await is a
  // boundary on both engines (V8 resolves it through an internal intrinsic,
  // not the then property), so async/await resumptions stay attributed to the
  // ambient turn.
  if (enabled()) {
    const originalThen = Promise.prototype.then;
    if (typeof originalThen === 'function') {
      const wrappedThen = function then(onFulfilled, onRejected) {
        return originalThen.call(
          this,
          G.__obscuraTraceBind(onFulfilled),
          G.__obscuraTraceBind(onRejected),
        );
      };
      try {
        Object.defineProperty(wrappedThen, 'length', {
          value: 2, writable: false, enumerable: false, configurable: true,
        });
        Object.defineProperty(wrappedThen, 'name', {
          value: 'then', writable: false, enumerable: false, configurable: true,
        });
      } catch (e) { /* descriptor best-effort */ }
      _markNative(wrappedThen);
      Promise.prototype.then = wrappedThen;
    }
  }

  publishFrom();
})();

File: crates/obscura-js/src/trace_source.rs
//! Execution-source attribution (`from` labels) for the trace streams.
//!
//! The label taxonomy mirrors the HaHaVM dispatch-trace `from` field so traces
//! from the two engines can be diffed by the code unit that produced each
//! record:
//!
//! | label | meaning |
//! | --- | --- |
//! | `window` | code in the main document's default scope |
//! | `iframe(N)` | code in a frame realm; N from the host's cross-frame counter |
//! | `worker(M)[creator]` | worker code; M from the host worker counter, creator from the constructing context's label at `new Worker(...)` |
//! | `script@<url>` | a classic script file (inline or external) |
//! | `function@<source>` | a `new Function(...)` product, attributed to its creation source |
//! | `eval@<source>` | eval'd string code (boundary: direct eval is not attributable in V8) |
//! | `host` | host-injected code (`Page::evaluate`, CDP evaluate, preloads, `--eval`) |
//!
//! The authoritative state lives here, thread-local in Rust: an ambient label
//! per context plus a stack of explicit entries. The JS bootstrap mirrors the
//! stack (it needs to read the current label for snapshotting continuations)
//! and keeps the two in lockstep through `op_trace_push_source` /
//! `op_trace_pop_source`. Rust-side funnels set the ambient label; host
//! injections push an explicit entry through [`push`].
//!
//! Everything is inert unless a trace stream is active
//! ([`enabled`]); with tracing off no label is written and no
//! wrapper is installed in the bootstrap, so the page-visible surface is
//! unchanged.

use std::cell::RefCell;

thread_local! {
    /// Ambient label of the context the runtime is about to execute in
    /// ("window", "iframe(N)", "worker(M)[...]"). Set by the Rust execution
    /// funnels, which know the target context.
    static AMBIENT: RefCell<String> = const { RefCell::new(String::new()) };
    /// Explicit entries pushed by labeled code units (script@..., host, ...).
    /// Innermost entry wins over the ambient label.
    static STACK: RefCell<Vec<String>> = const { RefCell::new(Vec::new()) };
}

/// Whether any trace stream requested execution-source labels. Mirrors the
/// runtime's snapshot-skip decision: a traced run re-executes the bootstrap
/// with `__obscura_trace_from_enabled` set, which is what installs the JS-side
/// wrappers.
///
/// The tracelog sink counts as a stream only when it was asked to stamp its
/// records with the label ([`crate::tracelog::labels_requested`]): without that
/// field the labels would add the wrapper cost to a run with nothing reading
/// them off.
pub fn enabled() -> bool {
    static ENABLED: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
    *ENABLED.get_or_init(|| {
        std::env::var_os("OBSCURA_TRACE_OP_FILE").is_some()
            || std::env::var_os("OBSCURA_TRACE_API_FILE").is_some()
            || crate::tracelog::labels_requested()
    })
}

/// The label a traced record in the currently-executing code should carry.
pub fn effective() -> String {
    let stacked = STACK.with(|stack| stack.borrow().last().cloned());
    match stacked {
        Some(label) => label,
        None => {
            let ambient = AMBIENT.with(|ambient| ambient.borrow().clone());
            if ambient.is_empty() {
                "window".to_string()
            } else {
                ambient
            }
        }
    }
}

/// Set the ambient label for the context the runtime is about to execute in.
/// Cheap no-op when tracing is off, which is the production path.
pub fn set_ambient(label: &str) {
    if !enabled() {
        return;
    }
    AMBIENT.with(|ambient| *ambient.borrow_mut() = label.to_string());
}

/// Restore point for [`push`]: pops the stack back to its prior depth, so a
/// leaked entry from a terminated script cannot outlive its bracket.
pub struct TraceSourceGuard {
    depth: usize,
}

impl Drop for TraceSourceGuard {
    fn drop(&mut self) {
        STACK.with(|stack| {
            let mut stack = stack.borrow_mut();
            stack.truncate(self.depth);
        });
    }
}

/// Push an explicit label for host-injected code. The returned guard pops the
/// stack back to its prior depth on drop, including any entries leaked by
/// nested JS that was terminated mid-script.
pub fn push(label: &str) -> TraceSourceGuard {
    let depth = STACK.with(|stack| {
        let mut stack = stack.borrow_mut();
        stack.push(label.to_string());
        stack.len() - 1
    });
    TraceSourceGuard { depth }
}

/// JS-side entry into a labeled code unit (script@..., continuation restore).
/// The bootstrap mirrors its own stack; this is the write side that keeps the
/// Rust stack authoritative for ops that trace.
pub(crate) fn js_push(label: &str) {
    if !enabled() {
        return;
    }
    STACK.with(|stack| stack.borrow_mut().push(label.to_string()));
}

/// JS-side exit from a labeled code unit.
pub(crate) fn js_pop() {
    if !enabled() {
        return;
    }
    STACK.with(|stack| stack.borrow_mut().pop());
}
